### DISCUSSION TITLE
Run multiple Procedures in sequence  / Nested Procedures

### DISCUSSION THREAD

**Discussion by bauzbue** [pymeasure/pymeasure#1094] [Q&A]
Hi all, 

What would be the best way to run multiple procedures in sequence? I would like to create multiple different procedures (and corresponding result files), and then run them one after the other without additional user input. 

I understand I could write a single procedure handling this, but I would really appreciate to have the logical granularity of each test-case seperated. I don't know what a proper way to achieve this may look like - Using the manager.queue to add new procedures? or is there a possibility to create nested procedures? A huge plus would be to keep the advantages of the Sequencer, Estimator and so on. 

I found a link here that prefers to have a single procedure in the end: https://github.com/pymeasure/pymeasure/discussions/700. 
And another guy, who apparently made a sequence as I would intend, but (i think) without all the nice GUI things and with no explanation on how: https://github.com/pymeasure/pymeasure/discussions/650

Any insights would be appreciated :)

Edit: Adding more details on my test-case (as an example): 

Testcase1: (Procedure1)
- Instruments: Signal Generator, Spectrum Analyzer, Supply, Thermo-Oven
- Measure: maximum output power with a sinusoidal input signal (over Settings [1,2,3], Supply-Voltages, Temperatue[-20, -10, 0, 10]...)
- Output: Results over said parameter in one file

Testcase2 : (Procedure2)
- Instruments: Same as above
-  Measure: Power of spurious signals with a two tone input signal (over **different** Settings [3,4,5,6], Supply Volt, Temp[+40, +80,+100],...)
- Output: Results over said parameter in one file

Testcase3 : (Procedure3)
- and so on 

With these test-cases I want to run a 'Test-Plan', somehing like: 

`for testcase in testPlan: `
`      run(testcase)`

**Comment by mcdo0486** [comment]
The short answer is, if you want to use the GUI, you'll need to pass in one `Procedure` class.  You mention the `Sequencer`, and it runs multiple conditions of the same procedure in sequence.

You could inherit properties and methods from other procedures into one combined procedure class, but the combined procedure would need one `execute()` section with all the logic.

Can you provide an example of the scenario you're thinking about with granular procedures and how you'd like to nest them?

**Reply by bauzbue** [reply]
Hi @mcdo0486 
I updated an example - I hope this makes it more clear what I would want to do. 

I think with one procedure class I may be limited in the functionality of the sequencer, as all inputs would have to be the same. 

I think this should be a very common problem, but I have not found a way to do this yet in pyMeasure - which suprises me.

**Reply by mcdo0486** [reply]
The great feature with the Sequencer is you can key off of the test plan condition and run different methods in your code depending on the condition. The sequencer will also create new result files for each run, since the Procedure gets shutdown and reinitialized. Something link this:

```
   selected_plan = Parameter("selected plan")
   settings = Parameters("overloaded settings")

    def execute(self):
        if self.selected_plan == '1':
            for i in range(self.iterations):
                # run plan1 that uses self.settings one way
                data = self.plan1_reading()
                self.emit(data)
                if self.should_stop():
                    log.warning("Catch stop command in procedure")
                    break
        elif self.selected_plan == '2':
            for i in range(self.iterations):
                # run plan2 that uses self.settings a different way
                data = self.plan2_reading()
                self.emit(data)
                if self.should_stop():
                    log.warning("Catch stop command in procedure")
                    break
```



Also checkout the "save sequence to file" feature. You can write a long, nested sequence file that can have all the conditions.

But if you really wanted "building block" Procedures that can be run independently or combined, you could do that by adding a new method that takes in the `self.emit()` and `self.should_stop()` functions as callbacks. That method would be called in its own `execute()` to run on its own, or it can be subclassed and called in another `execute()`. Kind of difficult to explain and it isn't perfect, but you can see an example in this gist. https://gist.github.com/mcdo0486/265bfa6669e23af8ab3b05ed9d2b4a6e

I wouldn't recommend this approach in the gist though. If you made significant changes to a "building block" procedure, you'd likely need to take that into account to any downstream procedure that inherits it, which defeats the convenience. It also wouldn't generate new files after each procedure runs. In my opinion, having procedure files be more explicit with what is happening is better in the long-term for experiment stability and sharing data/procedures. I'd recommend the Sequencer for your use case.
